core/src/main/java/com/linecorp/armeria/common/stream/ByteStreamMessageOutputStream.java

```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Override
+    public void abort() {
+        delegate.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        delegate.abort(cause);
+    }
+
+    private static final class OutputStreamSubscriber implements Subscriber<HttpData>, Subscription {
+
+        private final Subscriber<? super HttpData> downstream;
+        private final EventExecutor executor;
+
+        private final StreamMessageAndWriter<HttpData> outputStreamWriter;
+        private final Consumer<OutputStream> outputStreamConsumer;
+        private final Executor blockingTaskExecutor;
+
+        @Nullable
+        private Subscription upstream;
+        private boolean completed;
+
+        OutputStreamSubscriber(Subscriber<? super HttpData> downstream,
+                               EventExecutor executor,
+                               StreamMessageAndWriter<HttpData> outputStreamWriter,
+                               Consumer<OutputStream> outputStreamConsumer,
+                               Executor blockingTaskExecutor) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(executor, "executor");
+            requireNonNull(outputStreamWriter, "outputStreamWriter");
+            requireNonNull(outputStreamConsumer, "outputStreamConsumer");
+            requireNonNull(blockingTaskExecutor, "blockingTaskExecutor");
+            this.downstream = downstream;
+            this.executor = executor;
+            this.outputStreamWriter = outputStreamWriter;
+            this.outputStreamConsumer = outputStreamConsumer;
+            this.blockingTaskExecutor = blockingTaskExecutor;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            upstream = subscription;
+            downstream.onSubscribe(this);
+            blockingTaskExecutor.execute(() -> {
+                try {
+                    outputStreamConsumer.accept(new StreamWriterOutputStream(outputStreamWriter));
+                } catch (Throwable t) {
+                    onError(t);
+                    upstream.cancel();
+                }
+            });
+        }
+
+        @Override
+        public void onNext(HttpData data) {
+            requireNonNull(data, "data");
+            if (completed) {
+                data.close();
+                return;
+            }
+            downstream.onNext(data);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            requireNonNull(t, "t");
+            if (completed) {
+                return;
+            }
+            completed = true;
+            downstream.onError(t);
+        }
+
+        @Override
+        public void onComplete() {
+            if (completed) {
+                return;
+            }
+            completed = true;
+            downstream.onComplete();
+        }
+
+        @Override
+        public void request(long n) {
```

Comment:
Now we can remove `request()` and `cancel()`.

Reply:
✅ updated!